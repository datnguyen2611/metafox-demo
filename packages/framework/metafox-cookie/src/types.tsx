import { CookieAttributes } from 'js-cookie';

export interface CookieBackendConfig extends CookieAttributes {
  prefix: string;
  attributes?: CookieAttributes;
}
