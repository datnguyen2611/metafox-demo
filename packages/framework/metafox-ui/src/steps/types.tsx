import { RenderItemBase } from '@metafox/framework/types';

export interface StepShape extends RenderItemBase {
  title: string;
  id: string;
  expanded?: boolean;
  completed?: boolean;
  disabled?: boolean;
}

export interface State {
  activeStep: number;
  activeId?: string;
  data: { steps: StepShape[] };
}

export interface StepsProps {
  steps: StepShape[];
}

export interface ProcessStepShape {
  title: string;
  enableReport?: boolean;
  disableRetry?: boolean;
  dryRun?: boolean;
  disableUserAbort?: boolean;
  timeout?: number;
  dataSource: { apiUrl: string; apiMethod?: string };
  data?: object; // passed to api request for POST, PUT
  params?: object; // passed to api request for GET, DELETE
  message?: string;
}
export interface ProcessState {
  steps: StepShape[];
  current: number;
  done?: boolean;
}

export interface AppItemShape {
  identity: string;
  name: string;
  version: string;
}
