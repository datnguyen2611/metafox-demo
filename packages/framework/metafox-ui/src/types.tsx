import type {
  GridItemProps,
  HandleAction,
  RemoteDataSource
} from '@metafox/framework/types';
import { ThemeId } from '@metafox/layout';
import {
  ButtonProps,
  GridProps,
  IconButtonProps,
  PopperProps,
  SxProps,
  TypographyProps,
  TypographyVariant
} from '@mui/material';
import { DataGridProps } from '@mui/x-data-grid';
import { LinkProps } from 'react-router-dom';

export type GridColumns = DataGridProps['columns'];

export type ItemItemPropsShape = Record<string, any>;

export type ItemStatisticShape = {
  total_like: number;
  total_comment: number;
  total_score: number;
  total_rating: number;
  [key: string]: number;
  total_item: number;
  total_photo: number;
};

export interface FeedEmbedCardProps<A = any> {
  variant?: 'grid' | 'list' | 'default';
  bottomSpacing?: 'normal' | 'dense';
  children?: React.ReactNode;
  link?: string;
  item: A;
  feed: any;
  isShared?: boolean;
  sxOuter?: SxProps;
  sx?: SxProps;
}

export interface FeedEmbedCardMediaProps {
  image?: string;
  widthImage?: string;
  heightImage?: string;
  mediaRatio?: ImageRatio;
  variant?: 'grid' | 'list' | 'default';
  link?: string;
  playerOverlay?: boolean;
  playerOverlayProps?: TImagePlayerOverlay;
  host?: string;
}

export type TLineIconVariant = 'listItemIcon' | 'itemActionIcon';

export interface LineIconProps {
  icon: string;
  component?: React.ElementType;
  className?: string;
  variant?: TLineIconVariant;
  color?:
    | 'primary'
    | 'secondary'
    | 'textPrimary'
    | 'textSecondary'
    | 'inherit'
    | string;
  onClick?: (e?: any) => void;
  style?: any;
  role?: string;
  sx?: SxProps;
}

export type ItemExtraShape = {
  can_view: boolean;
  can_like: boolean;
  can_share: boolean;
  can_delete: boolean;
  can_delete_own: boolean;
  can_download: boolean;
  can_report: boolean;
  can_add: boolean;
  can_edit: boolean;
  can_comment: boolean;
  can_feature: boolean;
  can_approve: boolean;
  can_sponsor: boolean;
  can_sponsor_in_feed: boolean;
  can_purchase_sponsor: boolean;
  can_unlike: boolean;
  can_start_discussion: boolean;
  can_change_privacy_from_feed?: boolean;
  can_edit_cover_photo?: boolean;
  can_message?: boolean;
  can_upload_media?: boolean;
  can_add_cover?: boolean;
  can_remove_link_preview?: boolean;
  can_close?: boolean;
  can_show_payment_button?: boolean;
  can_show_no_payment_gateway_message?: boolean;
  can_view_comment?: boolean;
  can_show_sponsor_label?: boolean;
};

export type BackgroundSize = '48' | '300' | '1024' | 'origin';

export type AvatarSize = '50' | '120' | '200' | 'origin';

export type CoverSize =
  | '75'
  | '100'
  | '150'
  | '240'
  | '500'
  | '1024'
  | 'origin';

export interface ItemUserShape {
  resource_name: string;
  module_name: string;
  full_name: string;
  avatar: Record<AvatarSize, string>;
  link: string;
  is_featured: boolean;
  friend_id: number;
  user_name: string;
  id: number;
  cover: Record<CoverSize, string>;
  statistic: ItemStatisticShape;
  extra?: ItemExtraShape;
  city_location?: string;
  joined?: string;
}

export type ItemShape = {
  id: number | string;
  module_id: string;
  resource_name: string;
  is_featured: boolean;
  is_sponsor: boolean;
  is_sponsored_feed: boolean;
  is_liked: boolean;
  is_pending: boolean;
  is_online: boolean;
  item_id: number;
  creation_date: string;
  image?: Record<string, string>;
  images?: Record<string, string>[];
  extra?: ItemExtraShape;
  user: string;
  owner: string;
  price?: string;
  text_parsed?: string;
  statistic?: ItemStatisticShape;
  link?: string;
  url?: string;
  currency_id?: string;
  short_description?: string;
  description?: string;
  country?: string;
  city?: string;
  province?: string;
  privacy?: number;
  privacy_feed?: {
    icon: string;
    tooltip: string;
    label?: string;
  };
  privacy_detail?: {
    icon: string;
    tooltip: string;
    label?: string;
  };
  summary?: string;
  weblink?: string;
  attachments: string[];
  _identity: string;
};

export type SimpleUserShape = {
  id: number;
  user_name: string;
  full_name: string;
  avatar: string;
  friendship?: number;
  module_name?: string;
  link?: string;
};

export type ItemViewProps<T, U = unknown, S = unknown> = {
  identity: string;
  item: T;
  user?: SimpleUserShape & ItemUserShape;
  itemProps: ItemItemPropsShape;
  actions: U;
  state: S;
  setState: (cb: (prev: S) => S) => void;
  handleAction: (type: string, payload?: unknown) => void;
  wrapAs: React.FC<GridItemProps>;
  wrapProps: GridItemProps;
  pagingId?: string;
};

export type DetailViewProps<T, U = unknown, S = unknown> = {
  identity: string;
  item: T;
  user?: SimpleUserShape;
  actions: U;
  state: S;
  setState: (cb: (prev: S) => S) => void;
  handleAction: (type: string, payload?: unknown) => void;
};

export interface EmbedObjectShape {
  id: number | string;
  module_id: string;
  resource_name?: string;
  module_name?: string;
  is_featured: boolean;
  is_sponsor: boolean;
  is_liked: boolean;
  is_pending: boolean;
  item_id: number;
  creation_date: string;
  image?: string | Record<string, string>;
  images?: Record<string, string>;
  statistic: ItemStatisticShape;
  extra: ItemExtraShape;
  user: ItemUserShape;
  price?: string;
  link?: string;
  currency_id?: string;
  short_description?: string;
  country?: string;
  city?: string;
  province?: string;
  summary?: string;
  title?: string;
  description?: string;
}

export type EmbedItemInFeedItemProps<A = any> = {
  item: A;
  feed: any;
  isShared?: boolean;
  itemView?: string;
  photo: Record<string, any>;
};

export interface FeedEmbedObjectViewProps<
  A extends EmbedObjectShape = EmbedObjectShape,
  B = ItemShape
> {
  embed: A;
  feed: B;
  itemView?: string;
}

export type ItemEmbedObjectShape = ItemShape;

export interface UIBlockProps {
  variant?: string;
  noHeader?: boolean;
  noFooter?: boolean;
  [key: string]: any;
}

export interface UIGridContainerProps extends GridProps {}

export interface UIGridItemProps extends GridProps {}

export interface ItemTabProps {
  id?: string;
  tab: string;
  label: string;
  showWhen?: Record<string, any>;
  statisticKey?: string;
  [key: string]: any;
}

export interface UIBlockViewProps<A = RemoteDataSource, B = UIBlockProps> {
  title?: string;
  themeId?: ThemeId;
  tabProps?: {
    tabs: ItemTabProps[];
    activeTab: string;
    disableGutter: boolean;
    tabsNoSearchBox?: string[];
    placeholderSearch?: string;
    to?: string;
  };
  elements: any;
  dataSource?: A;
  blockView?: string;
  blockHeader?: string;
  blockFooter?: string;
  blockContent?: string;
  blockProps?: B;
}

export type ItemActionMenuProps = {
  id?: string;
  testid?: string;
  label?: string;
  identity?: string;
  menuName?: string;
  dependName?: string;
  appName?: string;
  icon?: string;
  keepMounted?: boolean;
  disableRipple?: boolean;
  className?: string;
  iconClassName?: string;
  children?: React.ReactNode;
  handleAction?: HandleAction;
  size?: ButtonProps['size'];
  color?: IconButtonProps['color'];
  variant?: IconButtonProps['variant'];
  // state?: any; // remove state pass from items.
  control?: JSX.Element;
  scrollRef?: React.RefObject<HTMLDivElement>;
  scrollClose?: boolean;
  placement?: PopperProps['placement'];
  items?: MenuItemShape[];
  disablePortal?: boolean;
  prefixName?: string;
  fallbackName?: string;
  tooltipTitle?: string;
  preventDefault?: boolean;
  sx?: SxProps;
  autoHide?: boolean;
  preventClose?: boolean;
  zIndex?: number;
  popperOptions?: any;
  popperStyles?: Record<string, any>;
  menuStyles?: Record<string, any>;
  variantPopper?: 'hidden-outview' | string;
  triggerOpen?: any;
};

export type MenuItemValue = string;

export interface MenuItemShape {
  className?: string;
  as?: 'link' | 'button' | 'divider' | 'popover' | 'heading' | string;
  note?: string;
  role?: string;
  target?: '_blank' | '_self' | '_parent' | '_top' | string;
  rel?: 'string';
  image?: string;
  dense?: boolean;
  alt?: string;
  active?: boolean;
  onClick?(): void;
  icon?: string;
  iconSecondary?: string;
  label?: string;
  labelSub?: string;
  value?: MenuItemValue;
  params?: Record<string, any>;
  showWhen?: any;
  enabledWhen?: any;
  variant?: 'text' | 'contained' | 'outlined';
  color?: string;
  disabled?: boolean;
  iconVariant?: any;
  testid?: string;
  to?: string;
  asModal?: boolean;
  asChildPage?: boolean;
  keepScroll?: boolean;
  item_name?: string;
  name?: string;
  buttonProps?: Record<string, any>;
  dividerProps?: {
    variant?: 'inset' | 'fullWidth' | 'middle';
    light?: boolean;
    className?: string;
  };
  placement?: PopperProps['placement'];
  items?: MenuItemShape[];
  order_by?: string;
  content?: {
    component: string | React.ElementType;
    props?: Record<string, any>;
  };
  tab?: string;
  preventClose?: boolean;
  dataSource?: { apiUrl?: string; apiParams?: string };
  behavior?: 'more' | 'last' | 'close';
  disablePortal?: boolean;
  menuItemType?: string;
}

export interface MenuItemViewProps {
  item: MenuItemShape;
  classes: Record<string, any>;
  handleAction?: HandleAction;
  pathname?: string;
  active?: boolean;
  handleCheckActiveMenu?: (url: string, path: string) => boolean;
  minimize?: boolean;
  isDesktop?: boolean;
  activeHover?: boolean;
  handlePopoverClose?: () => void;
  handlePopoverOpen?: () => void;
}

export interface MenuItemProps extends MenuItemShape {
  children?: JSX.Element;
  handleAction?: any;
  classes?: Record<
    'menuItem' | 'activeMenuItem' | 'menuLink' | 'menuItemIcon' | string,
    string
  >;
}

export type ImageRatio =
  | 'auto'
  | 'fixed'
  | '169'
  | '32'
  | '23'
  | '43'
  | '34'
  | '13'
  | '11';

export type ItemReactionShape = {
  resource_name: string;
  module_name: string;
  title: string;
  icon: string;
  color: string;
  ordering: number;
  is_default: boolean;
  id: number;
};

export type ItemCommentShape = {
  user: ItemUserShape;
  text: string;
  parent_id: string;
  extra: ItemExtraShape;
  creation_date: string;
  children?: string[];
  statistic: ItemStatisticShape;
  most_reactions?: string[];
  user_reacted?: string;
  extra_data?: Record<string, any>;
};

export type ActionCommentShape = {
  statistic: ItemStatisticShape;
  most_reactions?: Array<ItemReactionShape>;
};

export type TImagePlayerOverlay = {
  icon?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg';
};
export enum ImageMatureState {
  Normal = 0,
  Warning = 1,
  Strict = 2
}

export type ImageShape = {
  identity?: string;
  classes?: Record<string, any>;
  alt?: string;
  src: string;
  aspectRatio?: ImageRatio;
  shape?: 'square' | 'circle' | 'radius';
  link?: LinkProps['to'];
  host?: string;
  asModal?: boolean;
  backgroundImage?: boolean;
  playerOverlay?: boolean;
  imageFit?: 'cover' | 'contain';
  playerOverlayProps?: TImagePlayerOverlay;
  className?: string;
  imgClass?: string;
  draggable?: boolean;
  'data-testid'?: string;
  style?: Record<string, string>;
  mature?: ImageMatureState;
  linkParams?: Record<string, string>;
  matureProps?: Record<string, any>;
  onLoad?: () => void;
};

export interface TruncateTextProps extends TypographyProps {
  component?: React.ElementType;
  lines?: 0 | 1 | 2 | 3 | 4 | 5;
  fontWeight?: number;
  fixHeight?: boolean;
  showFull?: boolean;
  isIE?: boolean;
  notSupportLineClamp?: boolean;
  isSafari?: boolean;
}

export type ControlMenuItemProps = {
  item: MenuItemShape;
  handleAction?: HandleAction;
  variant?: TypographyVariant;
  as: React.ElementType;
  classes?: Record<string, string>;
  onClick?: () => void;
  className?: string;
  closeMenu?: () => void;
  active?: boolean;
};

export type TotalCommentProps = {
  total: number;
  handleAction?: HandleAction;
  message: string;
  identity: string;
  dataSource?: RemoteDataSource;
};
