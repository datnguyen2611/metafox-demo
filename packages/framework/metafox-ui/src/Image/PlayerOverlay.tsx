import { LineIcon } from '@metafox/ui';
import { styled } from '@mui/material';
import React from 'react';
import { TImagePlayerOverlay } from '../types';

const StyledOverlay = styled('div', {
  name: 'Overlay'
})<{ size: 'xs' | 'sm' | 'md' | 'lg' }>(({ size }) => ({
  position: 'absolute',
  left: 0,
  right: 0,
  top: 0,
  bottom: 0,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  pointerEvents: 'none',
  color: '#fff',
  ...(size === 'lg' && {
    fontSize: '96px'
  }),
  ...(size === 'md' && {
    fontSize: '72px'
  }),
  ...(size === 'sm' && {
    fontSize: '48px'
  }),
  ...(size === 'xs' && {
    fontSize: '24px'
  })
}));

const PlayerOverlay = ({
  icon,
  size = 'md',
  ...props
}: TImagePlayerOverlay) => {
  return (
    <StyledOverlay size={size} {...props}>
      <LineIcon icon={icon} />
    </StyledOverlay>
  );
};

export default PlayerOverlay;
