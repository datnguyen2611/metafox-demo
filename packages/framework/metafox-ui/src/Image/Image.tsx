/**
 * @type: ui
 * name: ui.image.default
 */
import { Link, useGlobal } from '@metafox/framework';
import { ImageShape, LineIcon, ImageMatureState } from '@metafox/ui';
import { styled, Theme, Box, Typography } from '@mui/material';
import { createStyles, makeStyles } from '@mui/styles';
import clsx from 'clsx';
import React from 'react';
import PlayerOverlay from './PlayerOverlay';
import { alpha } from '@mui/system/colorManipulator';

const name = 'Image';

const StyledBgImg = styled('span')({
  display: 'block',
  width: '100%',
  height: '100%',
  position: 'relative',
  backgroundPosition: 'center center',
  backgroundRepeat: 'no-repeat',
  backgroundOrigin: 'border-box',
  backgroundSize: 'cover'
});

const MatureImageItem = styled(Box, {
  name,
  slot: 'MatureImageItem',
  overridesResolver(props, styles) {
    return [styles.MatureImageItem];
  }
})<{ isStrict?: boolean }>(({ theme, isStrict }) => ({
  position: 'absolute',
  containerName: 'imageMature',
  containerType: 'inline-size',
  left: 0,
  right: 0,
  bottom: 0,
  top: 0,
  padding: theme.spacing(2),
  textAlign: 'center',
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
  width: '100%',
  cursor: 'pointer',
  ...(isStrict && {
    background: `radial-gradient(circle, ${alpha(
      theme.palette.grey[700],
      0.7
    )} 0%, ${alpha(theme.palette.grey[500], 0.5)} 100%)`
  })
}));
const InnerImage = styled(Box, {
  name,
  slot: 'InnerImage',
  overridesResolver(props, styles) {
    return [styles.InnerImage];
  }
})<{ isStrict?: boolean }>(({ theme }) => ({
  position: 'absolute',
  left: 0,
  right: 0,
  top: 0,
  bottom: 0,
  '& img': {
    width: '100%',
    height: '100%',
    maxWidth: '100%'
  }
}));

const DescriptionMature = styled(Typography, {
  name,
  slot: 'DescriptionMature',
  overridesResolver(props, styles) {
    return [styles.DescriptionMature];
  }
})(({ theme }) => ({
  userSelect: 'none',
  '@container imageMature (max-width: 250px)': {
    // hide description when image too small
    display: 'none'
  }
}));

const useStyles = makeStyles(
  (theme: Theme) =>
    createStyles({
      root: {
        display: 'block',
        position: 'relative',
        width: '100%',
        '& img': {
          maxWidth: '100%'
        },
        '&:before': {
          content: '""',
          display: 'block',
          paddingBottom: '100%'
        }
      },
      ratiofixed: {
        width: '100%',
        height: '100%',
        '&:before': {
          display: 'none'
        }
      },
      ratioauto: {
        '&:before': {
          display: 'none'
        },
        '& img': {
          maxHeight: '750px',
          width: '100%'
        }
      },
      ratio169: {
        '&:before': {
          paddingBottom: '56.25%'
        }
      },
      ratio32: {
        '&:before': {
          paddingBottom: '66.66%'
        }
      },
      ratio23: {
        '&:before': {
          paddingBottom: '150%'
        }
      },
      ratio43: {
        '&:before': {
          paddingBottom: '75%'
        }
      },
      ratio34: {
        '&:before': {
          paddingBottom: '133.33%'
        }
      },
      ratio13: {
        '&:before': {
          paddingBottom: '31%'
        }
      },
      ratio11: {
        '&:before': {
          paddingBottom: '100%'
        }
      },
      square: {
        borderRadius: '0',
        '& img': {
          borderRadius: '0'
        }
      },
      circle: {
        borderRadius: '100%',
        '& img': {
          borderRadius: '100%'
        }
      },
      radius: {
        borderRadius: theme.shape.borderRadius,
        '& img': {
          borderRadius: theme.shape.borderRadius
        }
      },
      inner: {
        position: 'absolute',
        left: 0,
        right: 0,
        top: 0,
        bottom: 0,
        '& img': {
          width: '100%',
          height: '100%',
          maxWidth: '100%'
        }
      },
      bgImage: {
        display: 'block',
        width: '100%',
        height: '100%',
        position: 'relative',
        backgroundPosition: 'center center',
        backgroundRepeat: 'no-repeat',
        backgroundOrigin: 'border-box'
      },
      blur1: {
        '& $bgImage': {
          filter: 'blur(25px)'
        },
        '& img': {
          filter: 'blur(25px)'
        }
      },
      blur2: {
        '& $bgImage': {
          filter: 'blur(5px)'
        },
        '& img': {
          filter: 'blur(5px)'
        }
      },
      cover: {
        '& $bgImage': {
          backgroundSize: 'cover'
        },
        '& img': {
          objectFit: 'cover'
        }
      },
      contain: {
        '& $bgImage': {
          backgroundSize: 'contain'
        },
        '& img': {
          objectFit: 'contain'
        }
      }
    }),
  { name: 'MuiImage' }
);

const DEFAULT_RATIO = 'auto';

const ImageBlock = ({
  aspectRatio = DEFAULT_RATIO,
  src,
  alt,
  imgClass,
  onLoad,
  backgroundImage,
  classes,
  playerOverlay,
  playerOverlayProps
}: ImageShape) => (
  <>
    {'auto' === aspectRatio ? (
      <img src={src} alt={alt} className={imgClass} onLoad={onLoad} />
    ) : (
      <InnerImage>
        {backgroundImage ? (
          <StyledBgImg
            className={classes.bgImage}
            style={{ backgroundImage: `url(${src})` }}
          />
        ) : (
          <img
            src={src}
            alt={alt}
            draggable={false}
            className={imgClass}
            onLoad={onLoad}
          />
        )}
      </InnerImage>
    )}
    {playerOverlay && <PlayerOverlay {...playerOverlayProps} />}
  </>
);

export default function Image(props: ImageShape) {
  const {
    aspectRatio = DEFAULT_RATIO,
    shape = 'square',
    link,
    linkParams: linkParamsProps = {},
    asModal,
    imageFit = 'cover',
    className,
    'data-testid': testid,
    style = {},
    matureProps = {},
    identity
  } = props;
  const maturePropsSetting = {
    color: 'white',
    showDescription: true,
    ...matureProps
  };
  const {
    dialogBackend,
    i18n,
    getAcl,
    dispatch,
    useGetItem,
    assetUrl,
    navigate
  } = useGlobal();
  const ageMature = getAcl('photo.photo.mature_photo_age_limit');
  const item = useGetItem(identity);

  const { mature: matureDefault, _mature, extra } = item || {};
  const mature = extra?.can_view_mature ? 0 : _mature ?? matureDefault ?? 0;
  const blurClass = mature !== 0 ? `blur${mature}` : undefined;

  const classes = useStyles();
  const strictImageBg = assetUrl('photo.album_no_image');
  const handleConfirmMature = React.useCallback(async () => {
    const ok = await dialogBackend.confirm({
      message: i18n.formatMessage({
        id: 'photo_mature_warning_desc'
      }),
      title: i18n.formatMessage({
        id: 'photo_mature_warning_title'
      })
    });

    if (ok) {
      if (identity) {
        dispatch({
          type: 'editItemLocal',
          payload: { identity, data: { mature: 0, _mature: 0 } }
        });
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const linkParams = { ...linkParamsProps, to: link };
  const handleClickStrictMature = React.useCallback(() => {
    navigate(link, { state: { asModal, ...linkParamsProps } });
  }, [navigate, link, linkParamsProps, asModal]);

  if (mature) {
    const isStrictMature = mature === ImageMatureState.Strict;

    return (
      <Box
        data-testid={testid}
        className={clsx(
          classes.root,
          classes[shape],
          classes[`ratio${aspectRatio}`],
          classes[imageFit],
          classes[blurClass],
          className
        )}
        sx={{
          ...style
        }}
        onClick={isStrictMature ? handleClickStrictMature : handleConfirmMature}
      >
        <ImageBlock
          {...props}
          src={isStrictMature ? strictImageBg : props?.src}
          classes={classes}
        />
        <MatureImageItem isStrict={isStrictMature} {...maturePropsSetting}>
          <LineIcon
            icon={isStrictMature ? 'ico-eye-off-o' : 'ico-warning-o'}
            sx={{ fontSize: '40px' }}
          />
          {maturePropsSetting?.showDescription ? (
            <DescriptionMature mt={0.5}>
              {i18n.formatMessage(
                {
                  id: isStrictMature
                    ? 'mature_photo_strict_message'
                    : 'mature_photo_warning_message'
                },
                { age: ageMature }
              )}
            </DescriptionMature>
          ) : null}
        </MatureImageItem>
      </Box>
    );
  }

  return link ? (
    <Link
      asModal={asModal}
      data-testid={testid}
      className={clsx(
        classes.root,
        classes[shape],
        classes[`ratio${aspectRatio}`],
        classes[imageFit],
        classes[blurClass],
        className
      )}
      style={style}
      {...linkParams}
    >
      <ImageBlock {...props} classes={classes} />
    </Link>
  ) : (
    <div
      data-testid={testid}
      className={clsx(
        classes.root,
        classes[shape],
        classes[`ratio${aspectRatio}`],
        classes[imageFit],
        classes[blurClass],
        className
      )}
      style={style}
    >
      <ImageBlock {...props} classes={classes} />
    </div>
  );
}
