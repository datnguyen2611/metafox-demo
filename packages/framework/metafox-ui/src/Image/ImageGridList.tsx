import { Box, ImageList, ImageListItem } from '@mui/material';
import React from 'react';
import Image from './Image';

const demoImage =
  'https://qctest.phpfox.us/CDN/file/pic/photo/2020/03/8f69134559b5c9763b2fb39f42e925e4_1024.jpg';

const demoData = [
  {
    img: demoImage,
    title: 'Image',
    author: 'author',
    cols: 1
  },
  {
    img: demoImage,
    title: 'Image2',
    author: 'author2',
    cols: 1
  }
];

export default function ImageGridList({ cols = 3, values = demoData }) {
  return (
    <Box
      sx={{
        display: 'flex',
        flexWrap: 'wrap',
        justifyContent: 'space-around',
        overflow: 'hidden'
      }}
    >
      <ImageList sx={{ width: '100%' }} cols={cols}>
        {values.map(tile => (
          <ImageListItem key={tile.img} cols={tile.cols || 1}>
            <Image src={tile.img} alt={tile.title} />
          </ImageListItem>
        ))}
      </ImageList>
    </Box>
  );
}
