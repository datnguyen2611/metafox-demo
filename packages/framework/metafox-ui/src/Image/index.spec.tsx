import { GlobalProvider, Manager } from '@metafox/framework';
import { ThemeProvider } from '@metafox/layout';
import { render } from '@testing-library/react';
import * as React from 'react';
import { MemoryRouter } from 'react-router';
import { ImageShape, TImagePlayerOverlay } from '../types';
import Image from './Image';
import PlayerOverlay from './PlayerOverlay';

const fallbackSrc = '/images/bg-img-1x1.png';
const MockInfoLoader = () => <div data-testid="infoLoader" />;

describe('Image', () => {
  const props: ImageShape = {
    'data-testid': 'image',
    alt: 'picture',
    src: fallbackSrc
  };
  const manager = Manager.factory({});
  const usePreference = jest.fn().mockImplementation(() => ({}));
  const getConfig = jest.fn().mockImplementation(() => {});
  const mockOn = jest.fn();
  const mockOff = jest.fn();

  manager.use({
    usePreference,
    eventCenter: { on: mockOn, off: mockOff },
    useLoggedIn: () => false,
    constants: { isLayoutPreviewWindow: false },
    getConfig
  });

  it('render', async () => {
    const { container } = render(
      <GlobalProvider value={manager}>
        <ThemeProvider infoLoader={MockInfoLoader}>
          <Image {...props} />
        </ThemeProvider>
      </GlobalProvider>
    );

    expect(container).toMatchSnapshot();
  });
  it('render ImageBlock aspectRatio:auto', async () => {
    const { container } = render(
      <GlobalProvider value={manager}>
        <ThemeProvider infoLoader={MockInfoLoader}>
          <MemoryRouter>
            <Image {...props} />
          </MemoryRouter>
        </ThemeProvider>
      </GlobalProvider>
    );

    expect(container).toMatchSnapshot();
  });
  it('render ImageBlock aspectRatio:169 playerOverlay', async () => {
    const propsItem: ImageShape = {
      ...props,
      aspectRatio: '169',
      playerOverlay: true,
      playerOverlayProps: {
        icon: 'ico-angle-right'
      }
    };
    const { container } = render(
      <GlobalProvider value={manager}>
        <ThemeProvider infoLoader={MockInfoLoader}>
          <MemoryRouter>
            <Image {...propsItem} />
          </MemoryRouter>
        </ThemeProvider>
      </GlobalProvider>
    );

    expect(container).toMatchSnapshot();
  });
  it('render ImageBlock aspectRatio:169 backgroundImage playerOverlay', async () => {
    const propsItem: ImageShape = {
      ...props,
      aspectRatio: '169',
      backgroundImage: true,
      playerOverlay: true,
      playerOverlayProps: {
        icon: 'ico-angle-right'
      }
    };
    const { container } = render(
      <GlobalProvider value={manager}>
        <ThemeProvider infoLoader={MockInfoLoader}>
          <MemoryRouter>
            <Image {...propsItem} />
          </MemoryRouter>
        </ThemeProvider>
      </GlobalProvider>
    );

    expect(container).toMatchSnapshot();
  });
  it('render link ', async () => {
    const propsItem = {
      ...props,
      link: '/home'
    };
    const { container } = render(
      <GlobalProvider value={manager}>
        <ThemeProvider infoLoader={MockInfoLoader}>
          <MemoryRouter>
            <Image {...propsItem} />
          </MemoryRouter>
        </ThemeProvider>
      </GlobalProvider>
    );

    expect(container).toMatchSnapshot();
  });
  it('render link host asModal', async () => {
    const propsItem = {
      ...props,
      host: 'host',
      asModal: true,
      link: '/home'
    };
    const { container } = render(
      <GlobalProvider value={manager}>
        <ThemeProvider infoLoader={MockInfoLoader}>
          <MemoryRouter>
            <Image {...propsItem} />
          </MemoryRouter>
        </ThemeProvider>
      </GlobalProvider>
    );

    expect(container).toMatchSnapshot();
  });
});

describe('PlayerOverlay', () => {
  it('render playerOverlay', async () => {
    const propsItem: TImagePlayerOverlay = {
      icon: 'ico-angle-right',
      size: 'lg'
    };
    const { container } = render(<PlayerOverlay {...propsItem} />);

    expect(container).toMatchSnapshot();
  });
});
