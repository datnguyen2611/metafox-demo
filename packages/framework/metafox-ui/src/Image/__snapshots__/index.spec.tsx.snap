// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image render 1`] = `
<div>
  <div
    data-testid="infoLoader"
  />
  <div
    class="MuiImage-root MuiImage-square MuiImage-ratioauto MuiImage-cover"
    data-testid="image"
  >
    <img
      alt="picture"
      src="/images/bg-img-1x1.png"
    />
  </div>
</div>
`;

exports[`Image render ImageBlock aspectRatio:169 backgroundImage playerOverlay 1`] = `
<div>
  <div
    data-testid="infoLoader"
  />
  <div
    class="MuiImage-root MuiImage-square MuiImage-ratio169 MuiImage-cover"
    data-testid="image"
  >
    <div
      class="MuiImage-inner"
    >
      <span
        class="ltr-1330e3l"
        style="background-image: url(/images/bg-img-1x1.png);"
      />
    </div>
    <div
      class="ltr-19ezo3w-Overlay-root"
    >
      <span
        class="ico ico-angle-right ltr-1o2x2pn-MuiIcon-root"
      />
    </div>
  </div>
</div>
`;

exports[`Image render ImageBlock aspectRatio:169 playerOverlay 1`] = `
<div>
  <div
    data-testid="infoLoader"
  />
  <div
    class="MuiImage-root MuiImage-square MuiImage-ratio169 MuiImage-cover"
    data-testid="image"
  >
    <div
      class="MuiImage-inner"
    >
      <img
        alt="picture"
        draggable="false"
        src="/images/bg-img-1x1.png"
      />
    </div>
    <div
      class="ltr-19ezo3w-Overlay-root"
    >
      <span
        class="ico ico-angle-right ltr-1o2x2pn-MuiIcon-root"
      />
    </div>
  </div>
</div>
`;

exports[`Image render ImageBlock aspectRatio:auto 1`] = `
<div>
  <div
    data-testid="infoLoader"
  />
  <div
    class="MuiImage-root MuiImage-square MuiImage-ratioauto MuiImage-cover"
    data-testid="image"
  >
    <img
      alt="picture"
      src="/images/bg-img-1x1.png"
    />
  </div>
</div>
`;

exports[`Image render link  1`] = `
<div>
  <div
    data-testid="infoLoader"
  />
  <a
    class="MuiImage-root MuiImage-square MuiImage-ratioauto MuiImage-cover"
    data-testid="image"
    href="/home"
  >
    <img
      alt="picture"
      src="/images/bg-img-1x1.png"
    />
  </a>
</div>
`;

exports[`Image render link host asModal 1`] = `
<div>
  <div
    data-testid="infoLoader"
  />
  <a
    class="MuiImage-root MuiImage-square MuiImage-ratioauto MuiImage-cover"
    data-testid="image"
    href="/home"
    target="_blank"
  >
    <img
      alt="picture"
      src="/images/bg-img-1x1.png"
    />
  </a>
</div>
`;

exports[`PlayerOverlay render playerOverlay 1`] = `
<div>
  <div
    class="css-3y3vt2-Overlay-root"
  >
    <span
      class="ico ico-angle-right css-1o2x2pn-MuiIcon-root"
    />
  </div>
</div>
`;
